module/os/linux/zfs/zpl_file.c

```diff
@@ -743,21 +746,46 @@ zpl_fallocate_common(struct inode *ip, int mode, loff_t offset, loff_t len)
 	spl_inode_lock(ip);
 	olen = i_size_read(ip);
 
-	if (offset > olen) {
-		spl_inode_unlock(ip);
-		return (0);
-	}
-	if (offset + len > olen)
-		len = olen - offset;
-	bf.l_type = F_WRLCK;
-	bf.l_whence = SEEK_SET;
-	bf.l_start = offset;
-	bf.l_len = len;
-	bf.l_pid = 0;
-
 	crhold(cr);
 	cookie = spl_fstrans_mark();
-	error = -zfs_space(ITOZ(ip), F_FREESP, &bf, O_RDWR, offset, cr);
+	if (mode & FALLOC_FL_PUNCH_HOLE) {
+		flock64_t bf;
+
+		if (offset > olen)
+			goto out_unmark;
+
+		if (offset + len > olen)
+			len = olen - offset;
+		bf.l_type = F_WRLCK;
+		bf.l_whence = SEEK_SET;
+		bf.l_start = offset;
+		bf.l_len = len;
+		bf.l_pid = 0;
+
+		error = -zfs_space(ITOZ(ip), F_FREESP, &bf, O_RDWR, offset, cr);
+	} else if ((mode & ~FALLOC_FL_KEEP_SIZE) == 0) {
```

Comment:
I'm having trouble wrapping my head around this. You've already ensured (via the check that returns EOPNOTSUPP) that nothing besides PUNCH_HOLE or KEEP_SIZE is set. This is the else arm, so we know that PUNCH_HOLE is not set. So at this point, KEEP_SIZE may or may not be set, but that is it. It looks like this condition is always true. I might be missing something, though, as I don't work with bitwise flags very often.

Reply:
Correct. If you look at your ABCD table, the if block covers C/D when PUNCH_HOLE is set, so this _has_ to be the A/B case.  It could have just been "else" right now, but I thought it was more forward looking to check for the `mode=0/2` case explicitly (file preallocation, without/with size change) to facilitate addition of more modes later in separate else blocks (or possibly a`switch`).  Even without the `-EOPNOTSUPP` check, we would know that this is `mode=0/2` since the `~KEEP_SIZE` mask would pass any other bits through and the result would not `== 0` in that case.

Reply:
FYI, for Boolean logic we learned [Karnaugh Maps](https://en.wikipedia.org/wiki/Karnaugh_map) in engineering to simplify the logic, which can be very helpful when working with multiple variables and trying to write a simplified expression to handle the cases.